File: Algorithms/714. Best Time to Buy and Sell Stock with Transaction Fee.rs
use std::cmp;

impl Solution {
    pub fn max_profit(prices: Vec<i32>, fee: i32) -> i32 {
        let mut cash: i32 = 0;
        let mut hold: i32 =  - prices[0];
        let length: usize = prices.len();
        
        for i in 0..length{
            cash = cmp::max(cash, hold + prices[i] - fee);
            hold = cmp::max(hold, cash - prices[i]);
        }
        
        cash
    }
}

File: Algorithms/714. Best Time to Buy and Sell Stock with Transaction Fee.ts
function maxProfit(prices: number[], fee: number): number {
  let result = 0;
  let hold = -prices[0];
  const n = prices.length;

  for (let i = 0; i < n; i++) {
    result = Math.max(result, hold + prices[i] - fee);
    hold = Math.max(hold, result - prices[i]);
  }

  return result;
}
